game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTerritoryValueUtils.java

```diff
@@ -429,18 +431,41 @@ private static double findWaterValue(
     return capitalOrFactoryValue / 100 + nearbyLandValue / 10;
   }
 
+  // 1700 oldT=4906 newT=227 newT2=166 newT3=152
+  @SuppressWarnings("ReferenceEquality")
   private static Set<Territory> findNearbyEnemyCapitalsAndFactories(
-      final Territory t, final Map<Territory, Double> enemyCapitalsAndFactoriesMap) {
-
-    Set<Territory> nearbyEnemyCapitalsAndFactories = new HashSet<>();
-    for (int i = MIN_FACTORY_CHECK_DISTANCE; i <= MAX_FACTORY_CHECK_DISTANCE; i++) {
-      nearbyEnemyCapitalsAndFactories = t.getData().getMap().getNeighbors(t, i);
-      nearbyEnemyCapitalsAndFactories.retainAll(enemyCapitalsAndFactoriesMap.keySet());
-      if (!nearbyEnemyCapitalsAndFactories.isEmpty()) {
-        break;
+      final Territory t, final Set<Territory> enemyCapitalsAndFactories) {
+    final GameMap map = t.getData().getMap();
+
+    int distance = 0;
+    Territory nextDistanceAfter = t;
+
+    final var found = new HashSet<Territory>();
+    final var visited = new HashSet<Territory>(List.of(t));
+    final var queue = new ArrayDeque<Territory>(List.of(t));
+    while (!queue.isEmpty()) {
+      final Territory node = queue.removeFirst();
+      for (final Territory t2 : map.getNeighbors(node)) {
+        if (visited.add(t2)) {
+          queue.add(t2);
+          if (enemyCapitalsAndFactories.contains(t2)) {
+            found.add(t2);
+            if (found.size() == enemyCapitalsAndFactories.size()) {
+              break;
+            }
+          }
+        }
       }
-    }
 
-    return nearbyEnemyCapitalsAndFactories;
+      if (node == nextDistanceAfter) {
```

Comment:
Even if it saves some redundant overhead, it is good practice to use equals over reference equality.
Have a look at the errorprone explanation for this "warning"/error: https://github.com/google/error-prone/blob/master/docs/bugpattern/ReferenceEquality.md

Reply:
The logic here is intended to compute distance without significant extra overhead. I'd have to check how much equals() adds. If it's non-trivial, might be better to just track distance explicitly (by having a Node type that includes a Territory and a distance member).

Reply:
I checked with .equals() and the result is a 20% slowdown. :(

Reply:
We added `ObjectUtils#referenceEquals()` some time ago to act as documentation for the case when one _really_ wants to use reference equality.  The Javadocs for this method state:

```java
* <p>Use this method only when you really need to compare object references for equality. In
* almost all cases you should be using {@link Object#equals(Object)}. <strong>Using this method
* documents your intention that a reference equality check is required.</strong>
```

Its usage should prevent ErrorProne from triggering on this line.

Reply:
Thanks, used that now.